docs/design.md

```diff
@@ -0,0 +1,371 @@
+   the full-shard index with no offset indication.
+   
+   These are calculated externally (e.g. semantic indexing service) and supplied
+   to the DAG store for storage and safekeeping.
+
+   A shard can have infinite number of semantic indices associated with it. Each
+   semantic index is identifid and stamped with its generation data (rule and
+   version).
+
+   We acknowledge that semantic indices perhaps don't belong in the DAG store
+   long-term. Despite that, we decide to incubate them here to potentially spin
+   them off in the future.
+
+3. **Top-level cross-shard index.** Aggregates of full shard indices that enable
+   shard routing of reads for concrete CIDs.
+
+### Interactions
+
+This component receives the queries coming in from the miner-side indexing
+sidecar, which in turn come from network indexers.
+
+It also serves the DAG access layer. When a shard is registered/acquired in the
+storage layer, and a Blockstore is demanded for it, the full index to provide
+random-access capability is obtained from the index repo.
+
+In the future, this subcomponent will also serve the top-level cross-shard
+index.
+
+### Interface
+
+```go
+type IndexRepo interface {
+   FullIndexRepo
+   ManifestRepo
+}
+
+type FullIndexRepo interface {
+   // public
+   GetFullIndex(key shard.Key) (FullIndex, error)
+
+   // private
+   InsertFullIndex(key shard.Key, index FullIndex) error
+   DeleteFullIndex(key shard.Key) (bool, error)
+}
+
+type ManifestRepo interface {
+   // public
+   GetManifest(key ManifestKey) (Manifest, error)
+   InsertManifest(key ManifestKey, manifest Manifest) error
+   DeleteManifest(key ManifestKey) (bool, error)
+   ExistsManifest(key ManifestKey) (bool, error)
+}
+
+type ManifestKey struct {
+   Shard    key.Shard
+   Rule     string
+   Version  string
+}
+
+type FullIndex interface {
+   Offset(c cid.Cid) (offset int64, err error) // inexistent: offset=-1, err=nil
+   Contains(c cid.Cid) (bool, error)
+   Len() (l int64, err error)
+   ForEach(func(c cid.Cid, offset int64) (ok bool, err error)) error
+}
+```
+
+## DAG access layer
+
+This layer is responsible for serving DAGs or sub-DAGs from one or many shards.
+Initially, queries will require the shard key. That is, queries will be bounded
+to a single identifiable shard.
+
+In the future, when the cross-shard top-level index is implemented, the DAG
+store will be capable of resolving the shard for any given root CID.
+
+DAG access layer allows various access abstractions:
+   1. Obtaining a Blockstore bound to a single shard.
+   2. Obtaining a Blockstore bound to a specific set of shards.
+   3. Obtaining a global Blockstore.
+   4. `ipld.Node` -- TBD.
+
+At this point, we are only concerned with (1). The remaining access patterns
+will be elaborated on in the future.
+
+```go
+type DAGAccessor interface {
+   Shard(key shard.Key) ShardAccessor
+}
+
+type ShardAccessor interface {
+   Blockstore() (ReadBlockstore, error)
+}
```

Comment:
We can add a `Close` here that calls `ReleaseShard` under the hood. It is the user's responsibility to call `Close` when it's done with the dag access.

Reply:
That sounds good to me, adding.